src/android/AuthStateManager.java

```diff
@@ -0,0 +1,168 @@
+            new AtomicReference<WeakReference<AuthStateManager>>(new WeakReference<AuthStateManager>(null));
+
+    private static final String TAG = "AuthStateManager";
+
+    private static final String STORE_NAME = "AuthState";
+    private static final String KEY_STATE = "state";
+
+    private final SharedPreferences mPrefs;
+    private final ReentrantLock mPrefsLock;
+    private final AtomicReference<AuthState> mCurrentAuthState;
+
+    @AnyThread
+    public static AuthStateManager getInstance(@NonNull Context context) {
+        AuthStateManager manager = INSTANCE_REF.get().get();
+        if (manager == null) {
+            manager = new AuthStateManager(context.getApplicationContext());
+            INSTANCE_REF.set(new WeakReference<AuthStateManager>(manager));
+        }
+
+        return manager;
+    }
+
+    private AuthStateManager(Context context) {
+        mPrefs = context.getSharedPreferences(STORE_NAME, Context.MODE_PRIVATE);
+        mPrefsLock = new ReentrantLock();
+        mCurrentAuthState = new AtomicReference<AuthState>();
+    }
+
+    @AnyThread
+    @NonNull
+    public AuthState getCurrent() {
+        if (mCurrentAuthState.get() != null) {
+            return mCurrentAuthState.get();
+        }
+
+        AuthState state = readState();
+        if (mCurrentAuthState.compareAndSet(null, state)) {
+            return state;
+        } else {
+            return mCurrentAuthState.get();
+        }
+    }
+
+    @AnyThread
+    @NonNull
+    public AuthState replace(@NonNull AuthState state) {
+        writeState(state);
+        mCurrentAuthState.set(state);
+        return state;
+    }
+
+    @AnyThread
+    @NonNull
+    public AuthState updateAfterAuthorization(
+            @Nullable AuthorizationResponse response,
+            @Nullable AuthorizationException ex) {
+        AuthState current = getCurrent();
+        current.update(response, ex);
+        return replace(current);
+    }
+
+    @AnyThread
+    @NonNull
+    public AuthState updateAfterTokenResponse(
+            @Nullable TokenResponse response,
+            @Nullable AuthorizationException ex) {
+        AuthState current = getCurrent();
+        current.update(response, ex);
+        return replace(current);
+    }
+
+    @AnyThread
+    @NonNull
+    public AuthState updateAfterRegistration(
+            RegistrationResponse response,
+            AuthorizationException ex) {
+        AuthState current = getCurrent();
+        if (ex != null) {
+            return current;
+        }
+
+        current.update(response);
+        return replace(current);
+    }
+
+    @AnyThread
+    @NonNull
+    private AuthState readState() {
+        mPrefsLock.lock();
+        try {
+            String currentState = mPrefs.getString(KEY_STATE, null);
+            if (currentState == null) {
+                return new AuthState();
+            }
+
+            try {
+                return AuthState.jsonDeserialize(currentState);
```

Comment:
The AppAuth docs say:
> Token requests, such as obtaining a new access token using a refresh token, follow a similar pattern: TokenRequest instances are dispatched using performTokenRequest() on an AuthorizationService instance, and a TokenResponse instance is returned via a callback.

But the only call to `performTokenRequest()` in this pull request is in `exchangeAuthorizationCode`.